### PR TITLE
DynamicTables component

### PR DESCRIPTION
**Pull request by FlxB2**
Adds DynamicTables component which contains functionality commonly used inside the winery when working with tables.
It enables developers to quickly add new tables with automatically generated Add/Edit/Remove modals.

The generation works by supplying WineryDynamicTableMetadata objects to the component. 
For example:

data: 
`{ key: 'someKey', type: 'float', description: 'test'}`
metadata:
```
[new DynamicTextData('key', 'Name', ...),
new DynamicDropdownData<'integer'|'float'>('type','Type',
      [{label: 'integer', value: 'integer'},
      {label: 'float', value: 'float'},), ], ...),
  new DynamicTextData('description','Description', ..)]
```
 
 Creates a table with the following rows: 'Name', 'Type', 'Description'. Furthermore, when clicking Add/Edit a modal is generated which contains two text fields and a dropdown form. The dropdown contains the options given in the constructor. Each metadata object is configurable by changing values in the constructor. It correlates to a component which contains the individual logic, templates etc.

- [x] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request. Especially, we require **a single commit**
- [x] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Ensure to use auto format in **all** files
- [x] Ensure that you appear in `NOTICE` at Copyright Holders
- [ ] Tests created for changes
- [x] Documentation updated (if needed)
- [ ] Screenshots added (for UI changes)

### CODE REVIEW

**Comment by genie-winery** [comment]
Can one of the admins verify this patch?

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/formComponents/dynamicDropdown.component.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    config: DynamicDropdownData;
+    group: FormGroup;
+}
+
+export class DynamicDropdownData<T = {}> extends WineryDynamicTableMetadata<T> {
+    controlType = 'dropdown';
+    validation: ValidatorFn[] | ValidatorFn;
+    options: { label: string, value: string }[] = [];
+
+    constructor( key: string,
```
Comment (miwurster) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/formComponents/dynamicDropdown.component.ts:40]: code format

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/wineryDynamicTable.service.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *******************************************************************************/
+
+import { Injectable } from '@angular/core';
+
+/**
+ * Creates grouped sets of input controls, based on the form model.
+ * It consumes the metadata from the {@link WineryDynamicFormData} model
+ */
+@Injectable()
+export class WineryDynamicTableService {
```
Comment (miwurster) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/wineryDynamicTable.service.ts:22]: Why is this needed?

Comment (FlxB2) [reply org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/wineryDynamicTable.service.ts:22]: Its not, I'll remove it!

File: org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/parameters/parameters.component.ts [hunk 3]
```diff
@@ -40,52 +43,89 @@ export class ParametersComponent {
...
+        this.dynamicTableData = [
+            new DynamicTextData(
+                'key',
+                'Name',
+                0,
+                [Validators.required],
+            )];
+        if (this.containsColumn('type')) {
+            this.dynamicTableData.push(
+                new DynamicDropdownData<'string' | 'integer' | 'float' | 'boolean' | 'timestamp'>(
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/parameters/parameters.component.ts:64]: Why don't you use the `YamlType` defined in the propertiesDefinition.component.ts? Maybe think about moving the types to a separate file.

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/wineryDynamicTableMetadata.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    controlType: string;
+
+    constructor(
+        key: string,
+        label?: string,
+        order?: number,
+        defaultValue?: T,
+        disabled?: boolean,
+        sortTableCol?: boolean) {
+        this.key = key;
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/wineryDynamicTableMetadata.ts:52]: You can shorten this whole class by definign the constructor like this (you can exchange private, protected or public here as you need it):

```ts
 constructor(
        private key: string,
        private label = '',
        private order = 1,
        private defaultValue?: T,
        private disabled = false,
        private sortTableCol = false) { }
```

This saves you lines 36 to 43 and 52 to 57.

File: org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/formComponents/dynamicDropdown.component.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                 label: string,
+                 options: { label: string, value: string }[],
+                 order?: number,
+                 defaultValue?: T,
+                 validation?: ValidatorFn[] | ValidatorFn,
+                 disabled?: boolean,
+                 sortTableCol?: boolean) {
+        super(key, label, order, defaultValue, disabled, sortTableCol);
+        this.options = options;
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/tosca-management/src/app/wineryDynamicTable/formComponents/dynamicDropdown.component.ts:49]: See the other comment at the constructor.

**Comment by lharzenetter** [comment]
And also make sure, that you have the `Signed-off-by: Felix Burk <felix.burk@googlemail.com>` in the squashed commit :)

File: org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/parameters/parameters.component.ts [hunk 6]
```diff
@@ -11,25 +11,28 @@
...
-    /* tslint:disable no-bitwise */
-    uuid: string = (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
+export class ParametersComponent implements OnInit {
 
     @Input() tableTitle = 'Parameters';
     @Input() modalTitle = 'Parameter';
+
     @Input() parameters: Parameter[] = [];
+
+    // TODO minimize this to some parameter object? no idea yet
```
[the remark this answers is not in the document]
Comment (FlxB2) [reply org.eclipse.winery.frontends/app/tosca-management/src/app/instance/sharedComponents/parameters/parameters.component.ts:35]: I actually decided to leave cloumns like this to prevent any issues with other components. I'll remove the comment

**Comment by miwurster** [comment]
@lharzenetter pls check and approve.

@FlxB2 pls squash and sign-off.

**Comment by miwurster** [comment]
@FlxB2 pls trigger and check the build again (we cannot merge until its green)

**Comment by lharzenetter** [comment]
LGTM, @FlxB2 can you please create a single commit? Afterward, we proceed with creating a CQ.

**Comment by miwurster** [comment]
Ping @FlxB2, pls resolve conflicts and squash the commits.

**Comment by miwurster** [comment]
CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22490

**Comment by lharzenetter** [comment]
Merged as of parallel checkin.
